File: internal/core/routing.go
package core

import (
	"cmp"
	"slices"
	"time"
)

type ScoringWeights struct {
	Reliability uint16 `json:"reliability"`
	Speed       uint16 `json:"speed"`
	Cost        uint16 `json:"cost"`
	FX          uint16 `json:"fx"`
	Liquidity   uint16 `json:"liquidity"`
	Operations  uint16 `json:"operations"`
}

func DefaultScoringWeights() ScoringWeights {
	return ScoringWeights{
		Reliability: 35,
		Speed:       20,
		Cost:        15,
		FX:          10,
		Liquidity:   10,
		Operations:  10,
	}
}

func (w ScoringWeights) Total() uint32 {
	return uint32(w.Reliability) +
		uint32(w.Speed) +
		uint32(w.Cost) +
		uint32(w.FX) +
		uint32(w.Liquidity) +
		uint32(w.Operations)
}

type BankPolicy struct {
	PolicyID             string              `json:"policy_id"`
	Version              uint32              `json:"version"`
	Weights              ScoringWeights      `json:"weights"`
	DisabledProviders    map[ProviderID]bool `json:"disabled_providers"`
	DisabledRoutes       map[RouteID]bool    `json:"disabled_routes"`
	MaxFXAgeSeconds      *uint64             `json:"max_fx_age_seconds,omitempty"`
	MinSuccessRateBps    *uint16             `json:"min_success_rate_bps,omitempty"`
	TargetP95LatencyMS   uint64              `json:"target_p95_latency_ms"`
	AutoSwitchingAllowed bool                `json:"auto_switching_allowed"`
}

func DefaultBankPolicy() BankPolicy {
	maxFXAge := uint64(180)
	minSuccess := uint16(8000)
	return BankPolicy{
		PolicyID:             "default",
		Version:              1,
		Weights:              DefaultScoringWeights(),
		DisabledProviders:    map[ProviderID]bool{},
		DisabledRoutes:       map[RouteID]bool{},
		MaxFXAgeSeconds:      &maxFXAge,
		MinSuccessRateBps:    &minSuccess,
		TargetP95LatencyMS:   120_000,
		AutoSwitchingAllowed: true,
	}
}

type EligibilityInput struct {
	TransactionID                  string           `json:"transaction_id"`
	Corridor                       Corridor         `json:"corridor"`
	PayoutMethod                   PayoutMethod     `json:"payout_method"`
	Amount                         Money            `json:"amount"`
	DestinationBank                string           `json:"destination_bank,omitempty"`
	ComplianceManualReviewRequired bool             `json:"compliance_manual_review_required"`
	CircuitBreakerBlockedRoutes    map[RouteID]bool `json:"circuit_breaker_blocked_routes"`
}

type RouteHealthSnapshot struct {
	SuccessRateBps        uint16    `json:"success_rate_bps"`
	UptimeBps             uint16    `json:"uptime_bps"`
	P95LatencyMS          uint64    `json:"p95_latency_ms"`
	ManualInterventionBps uint16    `json:"manual_intervention_bps"`
	SettlementReliability uint16    `json:"settlement_reliability_bps"`
	ObservedAt            time.Time `json:"observed_at"`
}

func DefaultRouteHealthSnapshot() RouteHealthSnapshot {
	return RouteHealthSnapshot{
		SuccessRateBps:        10_000,
		UptimeBps:             10_000,
		P95LatencyMS:          1_000,
		ManualInterventionBps: 0,
		SettlementReliability: 10_000,
		ObservedAt:            time.Now().UTC(),
	}
}

type RouteHealthBook struct {
	Routes map[RouteID]RouteHealthSnapshot `json:"routes"`
}

type RouteHealthProvider interface {
	SnapshotFor(routeID RouteID) RouteHealthSnapshot
}

func NewRouteHealthBook() RouteHealthBook {
	return RouteHealthBook{Routes: map[RouteID]RouteHealthSnapshot{}}
}

func (b RouteHealthBook) SnapshotFor(routeID RouteID) RouteHealthSnapshot {
	if snapshot, ok := b.Routes[routeID]; ok {
		return snapshot
	}
	return DefaultRouteHealthSnapshot()
}

type RejectionReason string

const (
	RejectProviderMissing        RejectionReason = "provider_missing"
	RejectProviderDisabled       RejectionReason = "provider_disabled"
	RejectProviderNotApproved    RejectionReason = "provider_not_approved"
	RejectRouteDisabled          RejectionReason = "route_disabled"
	RejectRouteBlocked           RejectionReason = "route_blocked"
	RejectRouteDegraded          RejectionReason = "route_degraded"
	RejectUnsupportedCorridor    RejectionReason = "unsupported_corridor"
	RejectUnsupportedPayout      RejectionReason = "unsupported_payout_method"
	RejectUnsupportedCurrency    RejectionReason = "unsupported_currency"
	RejectUnsupportedBank        RejectionReason = "unsupported_destination_bank"
	RejectAmountOutsideLimits    RejectionReason = "amount_outside_limits"
	RejectStaleFXRate            RejectionReason = "stale_fx_rate"
	RejectInsufficientLiquidity  RejectionReason = "insufficient_liquidity"
	RejectComplianceManualReview RejectionReason = "compliance_manual_review"
	RejectCircuitBreakerOpen     RejectionReason = "circuit_breaker_open"
	RejectBelowMinSuccessRate    RejectionReason = "below_minimum_success_rate"
)

type RejectedRoute struct {
	RouteID    RouteID         `json:"route_id"`
	ProviderID ProviderID      `json:"provider_id"`
	Reason     RejectionReason `json:"reason"`
}

type CandidateRouteScore struct {
	RouteID         RouteID           `json:"route_id"`
	ProviderID      ProviderID        `json:"provider_id"`
	TotalScoreBps   uint16            `json:"total_score_bps"`
	ComponentScores map[string]uint16 `json:"component_scores"`
	PrimaryReason   string            `json:"primary_reason"`
}

type RouteDecision struct {
	TransactionID        string                `json:"transaction_id"`
	PolicyID             string                `json:"policy_id"`
	PolicyVersion        uint32                `json:"policy_version"`
	SelectedRoute        *CandidateRouteScore  `json:"selected_route,omitempty"`
	EligibleRoutes       []CandidateRouteScore `json:"eligible_routes"`
	RejectedRoutes       []RejectedRoute       `json:"rejected_routes"`
	DecidedAt            time.Time             `json:"decided_at"`
	AutoSwitchingAllowed bool                  `json:"auto_switching_allowed"`
}

func SelectRoute(registry RouteRegistry, policy BankPolicy, health RouteHealthProvider, input EligibilityInput) RouteDecision {
	eligibleRoutes := make([]CandidateRouteScore, 0)
	rejectedRoutes := make([]RejectedRoute, 0)

	for _, route := range registry.Routes {
		if route.Corridor != input.Corridor {
			rejectedRoutes = append(rejectedRoutes, reject(route, RejectUnsupportedCorridor))
			continue
		}

		if reason, rejected := rejectionReason(route, registry, policy, health, input); rejected {
			rejectedRoutes = append(rejectedRoutes, reject(route, reason))
			continue
		}

		eligibleRoutes = append(eligibleRoutes, scoreRoute(route, policy, health.SnapshotFor(route.ID)))
	}

	slices.SortFunc(eligibleRoutes, func(left, right CandidateRouteScore) int {
		if scoreOrder := cmp.Compare(right.TotalScoreBps, left.TotalScoreBps); scoreOrder != 0 {
			return scoreOrder
		}
		return cmp.Compare(left.RouteID, right.RouteID)
	})

	var selected *CandidateRouteScore
	if len(eligibleRoutes) > 0 {
		value := eligibleRoutes[0]
		selected = &value
	}

	return RouteDecision{
		TransactionID:        input.TransactionID,
		PolicyID:             policy.PolicyID,
		PolicyVersion:        policy.Version,
		SelectedRoute:        selected,
		EligibleRoutes:       eligibleRoutes,
		RejectedRoutes:       rejectedRoutes,
		DecidedAt:            time.Now().UTC(),
		AutoSwitchingAllowed: policy.AutoSwitchingAllowed,
	}
}

func rejectionReason(route Route, registry RouteRegistry, policy BankPolicy, health RouteHealthProvider, input EligibilityInput) (RejectionReason, bool) {
	provider, ok := registry.Provider(route.ProviderID)
	if !ok {
		return RejectProviderMissing, true
	}

	if !provider.Enabled || policy.DisabledProviders[provider.ID] {
		return RejectProviderDisabled, true
	}

	if !provider.Approved {
		return RejectProviderNotApproved, true
	}

	if policy.DisabledRoutes[route.ID] {
		return RejectRouteDisabled, true
	}

	if input.CircuitBreakerBlockedRoutes[route.ID] {
		return RejectCircuitBreakerOpen, true
	}

	switch route.Status {
	case RouteEnabled, RouteRecoveryTesting:
	case RouteDisabled:
		return RejectRouteDisabled, true
	case RouteDegraded:
		return RejectRouteDegraded, true
	case RouteBlocked:
		return RejectRouteBlocked, true
	}

	if route.PayoutMethod != input.PayoutMethod {
		return RejectUnsupportedPayout, true
	}

	if route.SettlementCurrency != input.Amount.Currency {
		return RejectUnsupportedCurrency, true
	}

	if !route.AmountRange.Contains(input.Amount) {
		return RejectAmountOutsideLimits, true
	}

	if !route.SupportsDestinationBank(input.DestinationBank) {
		return RejectUnsupportedBank, true
	}

	if input.ComplianceManualReviewRequired {
		return RejectComplianceManualReview, true
	}

	if !route.LiquidityAvailable {
		return RejectInsufficientLiquidity, true
	}

	if policy.MaxFXAgeSeconds != nil {
		if route.FXRateAgeSeconds == nil || *route.FXRateAgeSeconds > *policy.MaxFXAgeSeconds {
			return RejectStaleFXRate, true
		}
	}

	if policy.MinSuccessRateBps != nil {
		if health.SnapshotFor(route.ID).SuccessRateBps < *policy.MinSuccessRateBps {
			return RejectBelowMinSuccessRate, true
		}
	}

	return "", false
}

func reject(route Route, reason RejectionReason) RejectedRoute {
	return RejectedRoute{
		RouteID:    route.ID,
		ProviderID: route.ProviderID,
		Reason:     reason,
	}
}

func scoreRoute(route Route, policy BankPolicy, health RouteHealthSnapshot) CandidateRouteScore {
	reliability := averageBps(health.SuccessRateBps, health.UptimeBps, health.SettlementReliability)
	speed := latencyScore(health.P95LatencyMS, policy.TargetP95LatencyMS)
	cost := uint16(10_000 - min(route.CostPenaltyBps, 10_000))
	fx := min(route.FXQualityBps, 10_000)
	liquidity := uint16(0)
	if route.LiquidityAvailable {
		liquidity = 10_000
	}
	operations := uint16(10_000 - min(health.ManualInterventionBps, 10_000))

	weights := policy.Weights
	totalWeight := max(weights.Total(), 1)
	total := (uint32(reliability)*uint32(weights.Reliability) +
		uint32(speed)*uint32(weights.Speed) +
		uint32(cost)*uint32(weights.Cost) +
		uint32(fx)*uint32(weights.FX) +
		uint32(liquidity)*uint32(weights.Liquidity) +
		uint32(operations)*uint32(weights.Operations)) / totalWeight

	return CandidateRouteScore{
		RouteID:       route.ID,
		ProviderID:    route.ProviderID,
		TotalScoreBps: uint16(total),
		ComponentScores: map[string]uint16{
			"reliability": reliability,
			"speed":       speed,
			"cost":        cost,
			"fx":          fx,
			"liquidity":   liquidity,
			"operations":  operations,
		},
		PrimaryReason: primaryReason(reliability, speed, cost, fx),
	}
}

func averageBps(values ...uint16) uint16 {
	sum := uint32(0)
	for _, value := range values {
		sum += uint32(value)
	}
	return uint16(sum / uint32(len(values)))
}

func latencyScore(p95LatencyMS, targetP95LatencyMS uint64) uint16 {
	if p95LatencyMS <= targetP95LatencyMS {
		return 10_000
	}

	overshoot := p95LatencyMS - targetP95LatencyMS
	penalty := uint16(min((overshoot*10_000)/max(targetP95LatencyMS, 1), 10_000))
	return 10_000 - penalty
}

func primaryReason(reliability, speed, cost, fx uint16) string {
	parts := []struct {
		name  string
		score uint16
	}{
		{name: "reliability", score: reliability},
		{name: "speed", score: speed},
		{name: "cost", score: cost},
		{name: "fx", score: fx},
	}
	slices.SortFunc(parts, func(left, right struct {
		name  string
		score uint16
	}) int {
		if scoreOrder := cmp.Compare(right.score, left.score); scoreOrder != 0 {
			return scoreOrder
		}
		return cmp.Compare(left.name, right.name)
	})
	return "strongest " + parts[0].name
}

File: crates/imsi-core/src/routing.rs
use crate::registry::{Corridor, Money, PayoutMethod, Route, RouteId, RouteRegistry, RouteStatus};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet};

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct ScoringWeights {
    pub reliability: u16,
    pub speed: u16,
    pub cost: u16,
    pub fx: u16,
    pub liquidity: u16,
    pub operations: u16,
}

impl Default for ScoringWeights {
    fn default() -> Self {
        Self {
            reliability: 35,
            speed: 20,
            cost: 15,
            fx: 10,
            liquidity: 10,
            operations: 10,
        }
    }
}

impl ScoringWeights {
    pub fn total(self) -> u32 {
        u32::from(self.reliability)
            + u32::from(self.speed)
            + u32::from(self.cost)
            + u32::from(self.fx)
            + u32::from(self.liquidity)
            + u32::from(self.operations)
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct BankPolicy {
    pub policy_id: String,
    pub version: u32,
    pub weights: ScoringWeights,
    pub disabled_providers: BTreeSet<String>,
    pub disabled_routes: BTreeSet<RouteId>,
    pub max_fx_age_seconds: Option<u64>,
    pub min_success_rate_bps: Option<u16>,
    pub target_p95_latency_ms: u64,
    pub auto_switching_allowed: bool,
}

impl Default for BankPolicy {
    fn default() -> Self {
        Self {
            policy_id: "default".into(),
            version: 1,
            weights: ScoringWeights::default(),
            disabled_providers: BTreeSet::new(),
            disabled_routes: BTreeSet::new(),
            max_fx_age_seconds: Some(180),
            min_success_rate_bps: Some(8_000),
            target_p95_latency_ms: 120_000,
            auto_switching_allowed: true,
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct EligibilityInput {
    pub transaction_id: String,
    pub corridor: Corridor,
    pub payout_method: PayoutMethod,
    pub amount: Money,
    pub destination_bank: Option<String>,
    pub compliance_manual_review_required: bool,
    pub circuit_breaker_blocked_routes: BTreeSet<RouteId>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct RouteHealthSnapshot {
    pub success_rate_bps: u16,
    pub uptime_bps: u16,
    pub p95_latency_ms: u64,
    pub manual_intervention_bps: u16,
    pub settlement_reliability_bps: u16,
    pub observed_at: DateTime<Utc>,
}

impl Default for RouteHealthSnapshot {
    fn default() -> Self {
        Self {
            success_rate_bps: 10_000,
            uptime_bps: 10_000,
            p95_latency_ms: 1_000,
            manual_intervention_bps: 0,
            settlement_reliability_bps: 10_000,
            observed_at: Utc::now(),
        }
    }
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct RouteHealthBook {
    pub routes: BTreeMap<RouteId, RouteHealthSnapshot>,
}

impl RouteHealthBook {
    pub fn snapshot_for(&self, route_id: &str) -> RouteHealthSnapshot {
        self.routes.get(route_id).cloned().unwrap_or_default()
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RejectionReason {
    ProviderMissing,
    ProviderDisabled,
    ProviderNotApproved,
    RouteDisabled,
    RouteBlocked,
    RouteDegraded,
    UnsupportedCorridor,
    UnsupportedPayoutMethod,
    UnsupportedCurrency,
    UnsupportedDestinationBank,
    AmountOutsideLimits,
    StaleFxRate,
    InsufficientLiquidity,
    ComplianceManualReview,
    CircuitBreakerOpen,
    BelowMinimumSuccessRate,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct RejectedRoute {
    pub route_id: RouteId,
    pub provider_id: String,
    pub reason: RejectionReason,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct CandidateRouteScore {
    pub route_id: RouteId,
    pub provider_id: String,
    /// 0-10000 where 10000 is best.
    pub total_score_bps: u16,
    pub component_scores: BTreeMap<String, u16>,
    pub primary_reason: String,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct RouteDecision {
    pub transaction_id: String,
    pub policy_id: String,
    pub policy_version: u32,
    pub selected_route: Option<CandidateRouteScore>,
    pub eligible_routes: Vec<CandidateRouteScore>,
    pub rejected_routes: Vec<RejectedRoute>,
    pub decided_at: DateTime<Utc>,
    pub auto_switching_allowed: bool,
}

pub fn select_route(
    registry: &RouteRegistry,
    policy: &BankPolicy,
    health: &RouteHealthBook,
    input: EligibilityInput,
) -> RouteDecision {
    let mut eligible_routes = Vec::new();
    let mut rejected_routes = Vec::new();

    for route in registry.routes.values() {
        if route.corridor != input.corridor {
            rejected_routes.push(reject(route, RejectionReason::UnsupportedCorridor));
            continue;
        }

        if let Some(reason) = rejection_reason(route, registry, policy, health, &input) {
            rejected_routes.push(reject(route, reason));
            continue;
        }

        eligible_routes.push(score_route(route, policy, &health.snapshot_for(&route.id)));
    }

    eligible_routes.sort_by(|left, right| {
        right
            .total_score_bps
            .cmp(&left.total_score_bps)
            .then_with(|| left.route_id.cmp(&right.route_id))
    });

    RouteDecision {
        transaction_id: input.transaction_id,
        policy_id: policy.policy_id.clone(),
        policy_version: policy.version,
        selected_route: eligible_routes.first().cloned(),
        eligible_routes,
        rejected_routes,
        decided_at: Utc::now(),
        auto_switching_allowed: policy.auto_switching_allowed,
    }
}

fn rejection_reason(
    route: &Route,
    registry: &RouteRegistry,
    policy: &BankPolicy,
    health: &RouteHealthBook,
    input: &EligibilityInput,
) -> Option<RejectionReason> {
    let provider = match registry.provider(&route.provider_id) {
        Some(provider) => provider,
        None => return Some(RejectionReason::ProviderMissing),
    };

    if !provider.enabled || policy.disabled_providers.contains(&provider.id) {
        return Some(RejectionReason::ProviderDisabled);
    }

    if !provider.approved {
        return Some(RejectionReason::ProviderNotApproved);
    }

    if policy.disabled_routes.contains(&route.id) {
        return Some(RejectionReason::RouteDisabled);
    }

    if input.circuit_breaker_blocked_routes.contains(&route.id) {
        return Some(RejectionReason::CircuitBreakerOpen);
    }

    match route.status {
        RouteStatus::Enabled | RouteStatus::RecoveryTesting => {}
        RouteStatus::Disabled => return Some(RejectionReason::RouteDisabled),
        RouteStatus::Degraded => return Some(RejectionReason::RouteDegraded),
        RouteStatus::Blocked => return Some(RejectionReason::RouteBlocked),
    }

    if route.payout_method != input.payout_method {
        return Some(RejectionReason::UnsupportedPayoutMethod);
    }

    if route.settlement_currency != input.amount.currency {
        return Some(RejectionReason::UnsupportedCurrency);
    }

    if !route.amount_range.contains(&input.amount) {
        return Some(RejectionReason::AmountOutsideLimits);
    }

    if !route.supports_destination_bank(input.destination_bank.as_deref()) {
        return Some(RejectionReason::UnsupportedDestinationBank);
    }

    if input.compliance_manual_review_required {
        return Some(RejectionReason::ComplianceManualReview);
    }

    if !route.liquidity_available {
        return Some(RejectionReason::InsufficientLiquidity);
    }

    if let Some(max_age) = policy.max_fx_age_seconds
        && route.fx_rate_age_seconds.is_none_or(|age| age > max_age)
    {
        return Some(RejectionReason::StaleFxRate);
    }

    if let Some(min_success) = policy.min_success_rate_bps {
        let snapshot = health.snapshot_for(&route.id);
        if snapshot.success_rate_bps < min_success {
            return Some(RejectionReason::BelowMinimumSuccessRate);
        }
    }

    None
}

fn reject(route: &Route, reason: RejectionReason) -> RejectedRoute {
    RejectedRoute {
        route_id: route.id.clone(),
        provider_id: route.provider_id.clone(),
        reason,
    }
}

fn score_route(
    route: &Route,
    policy: &BankPolicy,
    health: &RouteHealthSnapshot,
) -> CandidateRouteScore {
    let reliability = average_bps(&[
        health.success_rate_bps,
        health.uptime_bps,
        health.settlement_reliability_bps,
    ]);
    let speed = latency_score(health.p95_latency_ms, policy.target_p95_latency_ms);
    let cost = 10_000u16.saturating_sub(route.cost_penalty_bps.min(10_000));
    let fx = route.fx_quality_bps.min(10_000);
    let liquidity = if route.liquidity_available { 10_000 } else { 0 };
    let operations = 10_000u16.saturating_sub(health.manual_intervention_bps.min(10_000));

    let weights = policy.weights;
    let total_weight = weights.total().max(1);
    let total = (u32::from(reliability) * u32::from(weights.reliability)
        + u32::from(speed) * u32::from(weights.speed)
        + u32::from(cost) * u32::from(weights.cost)
        + u32::from(fx) * u32::from(weights.fx)
        + u32::from(liquidity) * u32::from(weights.liquidity)
        + u32::from(operations) * u32::from(weights.operations))
        / total_weight;

    let mut component_scores = BTreeMap::new();
    component_scores.insert("reliability".into(), reliability);
    component_scores.insert("speed".into(), speed);
    component_scores.insert("cost".into(), cost);
    component_scores.insert("fx".into(), fx);
    component_scores.insert("liquidity".into(), liquidity);
    component_scores.insert("operations".into(), operations);

    CandidateRouteScore {
        route_id: route.id.clone(),
        provider_id: route.provider_id.clone(),
        total_score_bps: total as u16,
        component_scores,
        primary_reason: primary_reason(reliability, speed, cost, fx),
    }
}

fn average_bps(values: &[u16]) -> u16 {
    let sum: u32 = values.iter().map(|value| u32::from(*value)).sum();
    (sum / values.len() as u32) as u16
}

fn latency_score(p95_latency_ms: u64, target_p95_latency_ms: u64) -> u16 {
    if p95_latency_ms <= target_p95_latency_ms {
        return 10_000;
    }

    let overshoot = p95_latency_ms - target_p95_latency_ms;
    let penalty = ((overshoot.saturating_mul(10_000)) / target_p95_latency_ms.max(1)) as u16;
    10_000u16.saturating_sub(penalty.min(10_000))
}

fn primary_reason(reliability: u16, speed: u16, cost: u16, fx: u16) -> String {
    let mut parts = [
        ("reliability", reliability),
        ("speed", speed),
        ("cost", cost),
        ("fx", fx),
    ];
    parts.sort_by(|left, right| right.1.cmp(&left.1).then_with(|| left.0.cmp(right.0)));
    format!("strongest {}", parts[0].0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::registry::{AmountRange, Provider};

    fn route(id: &str, provider_id: &str, cost_penalty_bps: u16) -> Route {
        Route {
            id: id.into(),
            provider_id: provider_id.into(),
            corridor: Corridor::new("US", "NG"),
            payout_method: PayoutMethod::BankAccount,
            amount_range: AmountRange {
                min_minor_units: 1,
                max_minor_units: 50_000_000,
            },
            settlement_currency: "NGN".into(),
            destination_banks: vec![],
            status: RouteStatus::Enabled,
            liquidity_available: true,
            fx_rate_age_seconds: Some(30),
            cost_penalty_bps,
            fx_quality_bps: 9_000,
        }
    }

    fn input() -> EligibilityInput {
        EligibilityInput {
            transaction_id: "txn-1".into(),
            corridor: Corridor::new("US", "NG"),
            payout_method: PayoutMethod::BankAccount,
            amount: Money::new("NGN", 10_000),
            destination_bank: Some("ACCESS".into()),
            compliance_manual_review_required: false,
            circuit_breaker_blocked_routes: BTreeSet::new(),
        }
    }

    #[test]
    fn selects_highest_scored_eligible_route() {
        let mut registry = RouteRegistry::default();
        registry.add_provider(Provider {
            id: "p1".into(),
            name: "Provider 1".into(),
            enabled: true,
            approved: true,
        });
        registry.add_provider(Provider {
            id: "p2".into(),
            name: "Provider 2".into(),
            enabled: true,
            approved: true,
        });
        registry.add_route(route("r1", "p1", 2_000));
        registry.add_route(route("r2", "p2", 0));

        let mut health = RouteHealthBook::default();
        health.routes.insert(
            "r1".into(),
            RouteHealthSnapshot {
                success_rate_bps: 8_200,
                p95_latency_ms: 80_000,
                ..RouteHealthSnapshot::default()
            },
        );
        health.routes.insert(
            "r2".into(),
            RouteHealthSnapshot {
                success_rate_bps: 9_800,
                p95_latency_ms: 30_000,
                ..RouteHealthSnapshot::default()
            },
        );

        let decision = select_route(&registry, &BankPolicy::default(), &health, input());

        assert_eq!(decision.selected_route.unwrap().route_id, "r2");
        assert_eq!(decision.eligible_routes.len(), 2);
    }

    #[test]
    fn rejects_stale_fx_and_open_circuit_breaker() {
        let mut registry = RouteRegistry::default();
        registry.add_provider(Provider {
            id: "p1".into(),
            name: "Provider 1".into(),
            enabled: true,
            approved: true,
        });
        let mut stale = route("stale", "p1", 0);
        stale.fx_rate_age_seconds = Some(999);
        registry.add_route(stale);
        registry.add_route(route("blocked", "p1", 0));

        let mut input = input();
        input
            .circuit_breaker_blocked_routes
            .insert("blocked".into());

        let decision = select_route(
            &registry,
            &BankPolicy::default(),
            &RouteHealthBook::default(),
            input,
        );

        assert!(decision.selected_route.is_none());
        assert_eq!(decision.rejected_routes.len(), 2);
        assert!(
            decision
                .rejected_routes
                .iter()
                .any(|route| route.reason == RejectionReason::StaleFxRate)
        );
        assert!(
            decision
                .rejected_routes
                .iter()
                .any(|route| route.reason == RejectionReason::CircuitBreakerOpen)
        );
    }
}
